### PR TITLE
Add mem disk fs

### PR DESCRIPTION
**Pull request by katcipis**

### CODE REVIEW

File: fs.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package gofs
+
+import "io"
+
+// ReaderFS is the interface that groups the basic Read file system methods.
```
Comment (i4ki) [inline fs.go:5]: very sane interfaces =) I wonder when operating systems will deliver APIs like that...

File: fs_test.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/NeowayLabs/gofs"
+)
+
+type fsBuilder func(t *testing.T) gofs.FS
+
+func testFS(t *testing.T, newfs fsBuilder) {
```
Comment (i4ki) [inline fs_test.go:17]: interesting the way tests are reused =)

File: fs_test.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+	assertEqualBytes(t, expectedContents, contents)
+}
+
+func testRemoveDir(t *testing.T, newfs fsBuilder) {
+	fs := newfs(t)
+	dir := newtestpath()
+	contents := []byte("echo")
+	file1 := dir + "/file1"
+	file2 := dir + "/file1"
```
Comment (i4ki) [inline fs_test.go]: typo?

File: memfs.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		if strings.HasPrefix(storedFile, dir) {
+			delete(m.fs, storedFile)
+			err = nil
+		}
+	}
+
+	return err
+}
+
+func (m *MemFS) lock() {
```
Comment (i4ki) [inline memfs.go:84]: avoiding embed the mutex to not export it?

**Review by i4ki — APPROVED** [review]
very nice implementaton and tests =)

File: fs.go (hunk 1 above)
Comment (katcipis) [reply fs.go:5]: a man can dream :D

File: fs_test.go (hunk 2 above)
Comment (katcipis) [reply fs_test.go:17]: Besides reuse I tried to capture parts of the protocol that Go interfaces cannot express but are part of the interface protocol. For example, io.Reader will return EOF on the final Read call, but this kind of thing can't be expressed on the interface definition, neither can it be validated by the compiler, but it is an essential part of any implementation of an io.Reader.

I had implemented similar things on the past on C++ using inheritance, it was much more chaotic and harder to understand, God bless Go =)

File: fs_test.go (hunk 3 above)
Comment (katcipis) [reply fs_test.go]: indeed, thanks for pointing it out

File: memfs.go (hunk 4 above)
Comment (katcipis) [reply memfs.go:84]: yep =P, I'm addicted to encapsulation, everything is private to me as default

File: fs_test.go (hunk 3 above)
Comment (katcipis) [reply fs_test.go]: fixed =)

**Pull request merged**